### ISSUE TITLE
no such file or directory, open '.../assets/dag-map.js.map'

### ISSUE THREAD

**Issue by stevenwu** [ember-cli/ember-cli-uglify#29]
I installed and setup for the first time. 
```
Build Canceled: Broccoli Builder ran into an error with UglifyWriter plugin. 💥
ENOENT: no such file or directory, open '/Users/steve/projects/marais/ui/tmp/uglify_writer-input_base_path-4C6JkBvo.tmp/assets/dag-map.js.map'
Error: ENOENT: no such file or directory, open '/Users/steve/projects/marais/ui/tmp/uglify_writer-input_base_path-4C6JkBvo.tmp/assets/dag-map.js.map'
```

**Comment by boris-petrov** [comment]
I have the same problem. Didn't happen with `2.0.0-beta.2`.

**Comment by Turbo87** [comment]
@stevenwu can you reproduce with a fresh `ember new` app? what steps are necessary?

**Comment by Turbo87** [comment]
this was likely caused by updating https://github.com/lydell/source-map-url from `0.3.0` to `0.4.0`, which I assume surfaced other issues in the libs that we're trying to minify. in other words: I don't think this is an issue with Uglify itself but rather in some lib that isn't distributing its sourcemap properly 🤔 

problem now is figuring out where `dag-map.js` comes from and why the `dag-map.js.map` file is missing even though it is probably announced to be available in `dag-map.js`.

**Comment by rwjblue** [comment]
dag-map is most likely krisselden/dag-map, but I don’t know why it would be manually concatted to the output (it isn’t part of a default ember-cli app)...

**Comment by Turbo87** [comment]
Not sure if it helps but I've hit this issue too now. With our app at Ember 2.12 everything builds as expected, but if I then upgrade the app to Ember 2.16 (both via bower, not npm) it fails with that error during prod build.

**Comment by rwjblue** [comment]
Ember internally uses dag-map perhaps it is including a sourcemap comment referencing the external file?

**Comment by stevenwu** [comment]
Yeah, I think the issue is sourcemap and one of the addons in my project.

**Comment by akatz** [comment]
It also only seems to hit if you build in prod mode.

**Comment by Turbo87** [comment]
It seems that this depends on the Ember version that you're running on. https://github.com/emberjs/ember.js/pull/15577 aimed to fix a similar issue, but it is not clear yet whether that fix also works for this issue here and in which Ember versions it was included.

**Comment by rwjblue** [comment]
@turbo87 - do you happen to know if this can be reproduced on a new app?

**Comment by Turbo87** [comment]
> do you happen to know if this can be reproduced on a new app?

I've just tried to reproduce this with a clean `ember new` app, but I'm not able to reproduce. Since this seems to work alright I don't think we should block the CLI release on it.

**Comment by boris-petrov** [comment]
@Turbo87, @rwjblue - here is a reproduction:

https://github.com/boris-petrov/ember-cli-uglify-issue

I did an `ember new`, use `ember` from `bower` (not sure if this matters) and added some of my packages in `package.json` (not sure which of those matter). If you run `ember build --environment=production` - it blows up. If you remove the `sourcemaps` property in `ember-cli-build.js` - it works fine.

I hope this is enough for you to debug it. Please tell me if I can help more.

**Comment by Turbo87** [comment]
@boris-petrov awesome, thanks. I think `ember` from `bower` might be the issue, because that is what I remember doing too when I had that issue... 🤔 

while this is certainly a valid issue, it is 1) an issue in Ember.js, not `ember-cli-uglify` and 2) an issue with Ember via bower which is no longer the default. so for those that have this issue you will have to stay on the old `ember-cli-uglify` version until this is resolved in Ember.js itself or migrate to using Ember via the `ember-source` npm dependency.

**Comment by boris-petrov** [comment]
@Turbo87 - I'm not sure what you mean that it is a problem in Ember.js. I've narrowed it down to `broccoli-uglify-sourcemap`'s usage of `source-map-url` - if I downgrade manually in `package-lock.json` the version of only that package from `0.4.0` to `0.3.0` and leave everything else as is (including `ember-cli-uglify` at `2.0.0`) - it works fine.

**Comment by Turbo87** [comment]
@boris-petrov the problem is that `source-map-url@0.4.0` fixed a bug, which uncovered another bug in how Ember.js is packaged up. So instead of reverting to `source-map-url@0.3.0` the proper solution is to fix the Ember.js packaging.

**Comment by boris-petrov** [comment]
I see. Ok, thanks, I'll be waiting for a fix then. Should I expect one soon or is this a low priority for you now?

**Comment by Turbo87** [comment]
> Should I expect one soon or is this a low priority for you now?

I'm not that involved with the Ember.js bundling process and got a few other things to wrap up first. Feel free to have a look :)

**Comment by akatz** [comment]
I'm working on an app that has migrated to ember-source and I can reproduce the error.

**Comment by akatz** [comment]
I have changed my package.json file to try and track down when this started happening. This is what I found, changing only the `ember-source` dependency. ( If you want the full yarn.lock file for each change let me know)
does not happen on 2.13.4
does happen on 2.14.0
does happen on 2.15.2
does happen on 2.16.0-beta.1

**Comment by rwjblue** [comment]
OK, so it is not related to bower, but also ember-source

**Comment by rwjblue** [comment]
I need to understand the bug that was fixed in `source-map-url` to know why this is wrong. 

<details>
<summary>If you take a look at [ember-source's source](https://unpkg.com/ember-source@2.14.0/dist/ember.debug.js) you can see that we do have `//# sourceMappingURL=dag-map.js.map` in the output (for `dag-map` module).</summary>

```js
enifed("dag-map", ["exports"], function (exports) {
    "use strict";

    /**
     * A topologically ordered map of key/value pairs with a simple API for adding constraints.
     *
     * Edges can forward reference keys that have not been added yet (the forward reference will
     * map the key to undefined).
     */
    var DAG = function () {
        function DAG() {
            this._vertices = new Vertices();
        }
        /**
         * Adds a key/value pair with dependencies on other key/value pairs.
         *
         * @public
         * @param key    The key of the vertex to be added.
         * @param value  The value of that vertex.
         * @param before A key or array of keys of the vertices that must
         *               be visited before this vertex.
         * @param after  An string or array of strings with the keys of the
         *               vertices that must be after this vertex is visited.
         */
        DAG.prototype.add = function (key, value, before, after) {
            if (!key) throw new Error('argument `key` is required');
            var vertices = this._vertices;
            var v = vertices.add(key);
            v.val = value;
            if (before) {
                if (typeof before === "string") {
                    vertices.addEdge(v, vertices.add(before));
                } else {
                    for (var i = 0; i < before.length; i++) {
                        vertices.addEdge(v, vertices.add(before[i]));
                    }
                }
            }
            if (after) {
                if (typeof after === "string") {
                    vertices.addEdge(vertices.add(after), v);
                } else {
                    for (var i = 0; i < after.length; i++) {
                        vertices.addEdge(vertices.add(after[i]), v);
                    }
                }
            }
        };
        /**
         * @deprecated please use add.
         */
        DAG.prototype.addEdges = function (key, value, before, after) {
            this.add(key, value, before, after);
        };
        /**
         * Visits key/value pairs in topological order.
         *
         * @public
         * @param callback The function to be invoked with each key/value.
         */
        DAG.prototype.each = function (callback) {
            this._vertices.walk(callback);
        };
        /**
         * @deprecated please use each.
         */
        DAG.prototype.topsort = function (callback) {
            this.each(callback);
        };
        return DAG;
    }();
    exports.default = DAG;

    /** @private */
    var Vertices = function () {
        function Vertices() {
            this.length = 0;
            this.stack = new IntStack();
            this.path = new IntStack();
            this.result = new IntStack();
        }
        Vertices.prototype.add = function (key) {
            if (!key) throw new Error("missing key");
            var l = this.length | 0;
            var vertex;
            for (var i = 0; i < l; i++) {
                vertex = this[i];
                if (vertex.key === key) return vertex;
            }
            this.length = l + 1;
            return this[l] = {
                idx: l,
                key: key,
                val: undefined,
                out: false,
                flag: false,
                length: 0
            };
        };
        Vertices.prototype.addEdge = function (v, w) {
            this.check(v, w.key);
            var l = w.length | 0;
            for (var i = 0; i < l; i++) {
                if (w[i] === v.idx) return;
            }
            w.length = l + 1;
            w[l] = v.idx;
            v.out = true;
        };
        Vertices.prototype.walk = function (cb) {
            this.reset();
            for (var i = 0; i < this.length; i++) {
                var vertex = this[i];
                if (vertex.out) continue;
                this.visit(vertex, "");
            }
            this.each(this.result, cb);
        };
        Vertices.prototype.check = function (v, w) {
            if (v.key === w) {
                throw new Error("cycle detected: " + w + " <- " + w);
            }
            // quick check
            if (v.length === 0) return;
            // shallow check
            for (var i = 0; i < v.length; i++) {
                var key = this[v[i]].key;
                if (key === w) {
                    throw new Error("cycle detected: " + w + " <- " + v.key + " <- " + w);
                }
            }
            // deep check
            this.reset();
            this.visit(v, w);
            if (this.path.length > 0) {
                var msg_1 = "cycle detected: " + w;
                this.each(this.path, function (key) {
                    msg_1 += " <- " + key;
                });
                throw new Error(msg_1);
            }
        };
        Vertices.prototype.reset = function () {
            this.stack.length = 0;
            this.path.length = 0;
            this.result.length = 0;
            for (var i = 0, l = this.length; i < l; i++) {
                this[i].flag = false;
            }
        };
        Vertices.prototype.visit = function (start, search) {
            var _a = this,
                stack = _a.stack,
                path = _a.path,
                result = _a.result;
            stack.push(start.idx);
            while (stack.length) {
                var index = stack.pop() | 0;
                if (index >= 0) {
                    // enter
                    var vertex = this[index];
                    if (vertex.flag) continue;
                    vertex.flag = true;
                    path.push(index);
                    if (search === vertex.key) break;
                    // push exit
                    stack.push(~index);
                    this.pushIncoming(vertex);
                } else {
                    // exit
                    path.pop();
                    result.push(~index);
                }
            }
        };
        Vertices.prototype.pushIncoming = function (incomming) {
            var stack = this.stack;
            for (var i = incomming.length - 1; i >= 0; i--) {
                var index = incomming[i];
                if (!this[index].flag) {
                    stack.push(index);
                }
            }
        };
        Vertices.prototype.each = function (indices, cb) {
            for (var i = 0, l = indices.length; i < l; i++) {
                var vertex = this[indices[i]];
                cb(vertex.key, vertex.val);
            }
        };
        return Vertices;
    }();
    /** @private */
    var IntStack = function () {
        function IntStack() {
            this.length = 0;
        }
        IntStack.prototype.push = function (n) {
            this[this.length++] = n | 0;
        };
        IntStack.prototype.pop = function () {
            return this[--this.length] | 0;
        };
        return IntStack;
    }();
    //# sourceMappingURL=dag-map.js.map
});
```
</details>
<br>
A few things strike me here:

1. I thought `sourceMappingURL` was only valid as the last line of a file?
2. We likely don't actually include a `dag-map.js.map` in the vendor tree from ember-source (I'd have to double check)

I suppose for now, we can just strip the comment to squelch this error...

**Comment by Turbo87** [comment]
> I thought sourceMappingURL was only valid as the last line of a file?

apparently not, or at least that was the "bug" that was fixed in `source-map-url@0.4.0`

> I suppose for now, we can just strip the comment to squelch this error...

that would be my recommendation too

**Comment by Turbo87** [comment]
awesome, thank you @rwjblue!

**Comment by boris-petrov** [comment]
@Turbo87, @rwjblue - do you still need the repo with the reproduction or I can delete it?

**Comment by Turbo87** [comment]
@boris-petrov no longer needed, now that https://github.com/emberjs/ember.js/pull/15720 is merged. thanks! :)

**Comment by urbany** [comment]
Hey everyone, I'm having a similar issue with the addon `ember-popper`.
Everything worked in ember-cli-uglify@1.2.0, but in ember-cli-uglify@2.0.0 I get an error.
It only occurs when I build for production with sourcemaps enabled: `ember build -e production`

```
Build failed.
The Broccoli Plugin: [Fingerprint] failed with:
Error: ENOENT: no such file or directory, open '/.../my-project/tmp/uglify_writer-input_base_path-K803YqEY.tmp/assets/popper.js.map'
```

ember-popper seems to be exporting the `popper.js.map` file:
https://github.com/kybishop/ember-popper/blob/v0.6.4/index.js#L22

And [popper.js](https://github.com/FezVrasta/popper.js/blob/v1.12.5/dist/umd/popper.js#L2448) has this: 
```
//# sourceMappingURL=popper.js.map
```

You can test with this repo: https://github.com/urbany/ember-popper-sourcemaps-error
`ember build -e production`

EDIT:
After a closer look I can see that in `/.../my-project/tmp/uglify_writer-input_base_path-K803YqEY.tmp/assets/vendor.js` still has this: ```//# sourceMappingURL=popper.js.map```, this should probably be stripped sinde the `popper.js.map`is already concatenated in the vendor.map

**Comment by maxvoltage** [comment]
I encountered this issue with swiper module, which has sourceMappingURL at the end of its minified js file in its dist folder. Will the solution be removing sourceMappingURL comment from it? Should I do the same for any other module I encounter to have a correct vendor.js.map?
I am on ember-source 2.16.2
also reproducible in 2.18

I opened a [PR](https://github.com/ember-cli/broccoli-uglify-sourcemap/pull/58) with broccoli-uglify-sourcemap. vendor.js file can have multiple sourceMappingURL locations when modules have their own sourceMappingURL. source-map-url module finds only one sourceMappingURL at a time. This PR removes sourceMappingURL if it doesn't have the same name with the js file until it finds the correct one.
